opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java

```diff
@@ -1109,66 +1116,68 @@ private static String readOpening(InputStream in, byte[] sig)
+        var opening = new StringBuilder();
+        try {
+            var readr = new BufferedReader(new InputStreamReader(in, encoding), OPENING_MAX_CHARS);
+            var ignoreWhiteSpace = true;
+            var breakOnNewLine = false;
+            for (int nRead = 0, r = readr.read(); r != -1 && nRead <= OPENING_MAX_CHARS;
+                 r = readr.read(), nRead++) {
+
+                var c = (char) r;
+                if (breakOnNewLine && c == '\n') {
+                    break;
                 }
-                sawNonWhitespace = true;
-            }
-            if (c == '\n') {
-                break;
-            }
+                if (Character.isWhitespace(c)) {
+                    if (!ignoreWhiteSpace) {
+                        opening.append(' ');
+                        ignoreWhiteSpace = true;
+                    }
+                } else {
+                    opening.append(c);
+                    // If the opening starts with "#!", then track so that any
+                    // trailing whitespace after the hashbang is ignored.
+                    ignoreWhiteSpace = opening.length() == 2 && opening.charAt(0) == '#' && opening.charAt(1) == '!';
+                    breakOnNewLine = true;
 
-            if (isWhitespace) {
-                // Track `lastWhitespace' to condense stretches of whitespace,
-                // and use ' ' regardless of actual whitespace character to
-                // accord with magic string definitions.
-                if (!lastWhitespace && !postHashbang) {
-                    opening.append(' ');
                 }
-            } else {
-                opening.append(c);
-                postHashbang = false;
-            }
-            lastWhitespace = isWhitespace;
-
-            // If the opening starts with "#!", then track so that any
-            // trailing whitespace after the hashbang is ignored.
-            if (opening.length() == 2 && opening.charAt(0) == '#' && opening.charAt(1) == '!') {
-                postHashbang = true;
             }
+        } catch (IOException e) {
+            throw new WrapperIOException(e);
         }
 
-        in.reset();
         return opening.toString();
     }
 
+    /**
+     * Identify the BOM encoding or default to utf 8, skips the encoding bytes .
+     *  Empty optional is returned if all bom bytes are not skipped
+     * @param in The input stream containing the data
+     * @param sig The initial sequence of bytes in the input stream
+     * @return optional of encoding
+     * @throws IOException in case of any read error
+     */
+    private static Optional<String> findAndSkipBomEncoding(InputStream in, byte[] sig) throws IOException {
+        var encoding = IOUtils.findBOMEncoding(sig);
+        if (Objects.isNull(encoding)) {
+            // SRCROOT is read with UTF-8 as a default.
+            encoding = StandardCharsets.UTF_8.name();
+        }
+        var skipForBOM = IOUtils.skipForBOM(sig);
```

Comment:
`skipForBOM` will call `findBOMEncoding()` once again for the same input. Perhaps it should have a variant where `encoding` can be passed in. Not this PR, though.

Reply:
Will take this up as a separate PR